File: Tree/Lowest Common Ancestor/3584번 - 가장 가까운 공통 조상/3584.py
import sys

input = sys.stdin.readline

class testCase():
    def __init__(self, a, b, p):
        self._a = a
        self._b = b
        self._p = p
        
    def lca(self):
        vis = [False] * len(self._p)
        vis[self._a] = True
        
        while self._a != self._p[self._a]:
            self._a = self._p[self._a]
            vis[self._a] = True
        while not vis[self._b]:
            self._b = self._p[self._b]
            
        return self._b

def solve():
    tc = int(input())
    for _ in range(tc):
        n = int(input())
        parents = list(range(n + 1))
        
        for _ in range(n - 1):
            a, b = map(int, input().split())
            parents[b] = a;
        
        fd_a, fd_b = map(int, input().split())
        print(testCase(fd_a, fd_b, parents).lca())
    
if __name__ == '__main__':
    solve()

File: Tree/Lowest Common Ancestor/3584번 - 가장 가까운 공통 조상/3584.cpp
#include <iostream>
#include <vector>

#define MAX 10001

using namespace std;

int t, n, m;
int parents[MAX];
vector<int> v[MAX];

void input() {
    for(int i = 0; i < MAX; i++) v[i].clear();
    for(int i = 1; i <= n; i++) parents[i] = i;
    
    cin >> n;
    for(int i = 0, a, b; i < n - 1; i++) {
        cin >> a >> b;
        parents[b] = a;
    }
}

int lca(int x, int y) {
    bool vis[MAX] = { false, };
    vis[x] = true;
    
    while(x != parents[x]) {
        x = parents[x];
        vis[x] = true;
    }
    while(!vis[y]) {
        y = parents[y];
    }
    
    return y;
}

void solve() {
    int u, v;
    cin >> u >> v;
    
    cout << lca(u, v) << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    cin >> t;
    while(t--) {
        input();
        solve();
    }
    
    return 0;
}
